platform/linux-generic/arch/aarch64/odp_crypto_armv8.c

```diff
@@ -137,6 +154,280 @@ null_crypto_routine(odp_packet_t pkt ODP_UNUSED,
 	return ODP_CRYPTO_ALG_ERR_NONE;
 }
 
+static
+odp_crypto_alg_err_t aes_gcm_encrypt(odp_packet_t pkt,
+				     const odp_crypto_packet_op_param_t *param,
+				     odp_crypto_generic_session_t *session)
+{
+	armv8_cipher_state_t cs = {
+		.counter = {
+			.d = {0, 0}
+		}
+	};
+	armv8_cipher_constants_t cc = {
+		.mode = 0
+	};
+	uint8_t *iv_ptr;
+	uint64_t nonce_length, nonce_byte_length;
+	uint64_t plaintext_length, plaintext_byte_length;
+	uint64_t aad_length, aad_byte_length;
+	int ret = 0;
+
+	/* Clear the allocated buffers */
+	memset(global->input, 0, _ARM_CRYPTO_MAX_DATA_LENGTH);
+	memset(global->nonce, 0, _ARM_CRYPTO_MAX_IV_LENGTH);
+	memset(global->aad, 0, _ARM_CRYPTO_MAX_AAD_LENGTH);
+	memset(global->digest, 0, _ARM_CRYPTO_MAX_DIGEST_LENGTH);
+	memset(global->output, 0, (_ARM_CRYPTO_MAX_DATA_LENGTH +
+	       _ARM_CRYPTO_MAX_DIGEST_LENGTH));
+
+	nonce_length = (session->p.cipher_iv.length) * 8;
+	/* IV length must be a multiple of 16 */
+	nonce_byte_length = ((nonce_length + 127) & ~127UL) >> 3;
+
+	if (param->cipher_iv_ptr)
+		iv_ptr = param->cipher_iv_ptr;
+	else if (session->p.cipher_iv.data)
+		iv_ptr = session->cipher.iv_data;
+	else
+		return ODP_CRYPTO_ALG_ERR_IV_INVALID;
+
+	/* Copy nonce and pad it to nonce_byte_length */
+	for (uint64_t i = 0; i < session->p.cipher_iv.length; i++)
+		global->nonce[i] = iv_ptr[i];
+	for (uint64_t i = session->p.cipher_iv.length; i < nonce_byte_length; i++)
+		global->nonce[i] = 0xde;
+
+	plaintext_length = (param->cipher_range.length) * 8;
+	/* Plaintext length must be a multiple of 16 */
+	plaintext_byte_length = ((plaintext_length + 127) & ~127UL) >> 3;
+
+	/* Copy plaintext and pad it to plaintext_byte_length */
+	uint32_t offset = param->cipher_range.offset;
+	uint32_t seg_len = 0;
+	uint8_t *data = odp_packet_offset(pkt, offset, &seg_len, NULL);
+
+	for (uint64_t i = 0; i < param->cipher_range.length; i++)
+		global->input[i] = data[i];
+	for (uint64_t i = param->cipher_range.length; i < plaintext_byte_length; i++)
+		global->input[i] = 0xde;
```

Comment:
What is 0xde? The last plain text block should be padded with zeroes, but I would suppose the crypto library takes care of it so that this is not needed at all (?).

Reply:
Answered above